crates/loro-internal/src/encoding/json_schema.rs

```diff
@@ -0,0 +1,1144 @@
+        pub value: FutureOp,
+        pub prop: i32,
+    }
+
+    #[derive(Debug, Clone, Serialize, Deserialize)]
+    #[serde(tag = "type", rename_all = "snake_case")]
+    pub enum ListOp {
+        Insert {
+            pos: usize,
+            value: LoroValue,
+        },
+        Delete {
+            pos: isize,
+            len: isize,
+            #[serde(with = "self::serde_impl::id")]
+            start_id: ID,
+        },
+    }
+
+    #[derive(Debug, Clone, Serialize, Deserialize)]
+    #[serde(tag = "type", rename_all = "snake_case")]
+    pub enum MovableListOp {
+        Insert {
+            pos: usize,
+            value: LoroValue,
+        },
+        Delete {
+            pos: isize,
+            len: isize,
+            #[serde(with = "self::serde_impl::id")]
+            start_id: ID,
+        },
+        Move {
+            from: u32,
+            to: u32,
+            #[serde(with = "self::serde_impl::idlp")]
+            elem_id: IdLp,
+        },
+        Set {
+            #[serde(with = "self::serde_impl::idlp")]
+            elem_id: IdLp,
+            value: LoroValue,
+        },
+    }
+
+    #[derive(Debug, Clone, Serialize, Deserialize)]
+    #[serde(tag = "type", rename_all = "snake_case")]
+    pub enum MapOp {
+        Insert { key: String, value: LoroValue },
+        Delete { key: String },
+    }
+
+    #[derive(Debug, Clone, Serialize, Deserialize)]
+    #[serde(tag = "type", rename_all = "snake_case")]
+    pub enum TextOp {
+        Insert {
+            pos: u32,
+            text: String,
+        },
+        Delete {
+            pos: isize,
+            len: isize,
+            #[serde(with = "self::serde_impl::id")]
+            id_start: ID,
+        },
+        Mark {
+            start: u32,
+            end: u32,
+            style_key: String,
+            style_value: LoroValue,
+            info: u8,
+        },
+        MarkEnd,
+    }
+
+    #[derive(Debug, Clone, Serialize, Deserialize)]
+    #[serde(tag = "type", rename_all = "snake_case")]
+    pub enum TreeOp {
+        Create {
+            #[serde(with = "self::serde_impl::tree_id")]
+            target: TreeID,
+            #[serde(with = "self::serde_impl::option_tree_id")]
+            parent: Option<TreeID>,
+            #[serde(default)]
+            fractional_index: FractionalIndex,
+        },
+        Move {
+            #[serde(with = "self::serde_impl::tree_id")]
+            target: TreeID,
+            #[serde(with = "self::serde_impl::option_tree_id")]
+            parent: Option<TreeID>,
+            #[serde(default)]
+            fractional_index: FractionalIndex,
+        },
+        Delete {
+            #[serde(with = "self::serde_impl::tree_id")]
+            target: TreeID,
+        },
+    }
+
+    #[derive(Debug, Clone, Serialize)]
+    #[serde(tag = "type", rename_all = "snake_case")]
+    pub enum FutureOp {
+        #[cfg(feature = "counter")]
+        Counter(OwnedValue),
+        Unknown(OwnedValue),
```

Comment:
Doesn't seem to align with the ts type definition?

Reply:
I think it's aligned. Do you mean there should be a extra wrapper？ I set it untagged

Reply:
I think it's aligned. This wrapper is untagged